src/lib/communication/DY.Lib.Communication.Core.Lemmas.fst

```diff
@@ -0,0 +1,573 @@
+  tr:trace ->
+  higher_layer_preds:comm_higher_layer_event_preds a ->
+  comm_keys_ids:communication_keys_sess_ids ->
+  sender:principal -> receiver:principal -> payload:a ->
+  Lemma
+  (requires
+    trace_invariant tr /\
+    has_private_keys_invariant /\
+    has_communication_layer_crypto_predicates /\
+    has_communication_layer_event_predicates higher_layer_preds /\
+    higher_layer_preds.send_auth tr sender payload /\
+    is_well_formed a (is_publishable tr) payload
+  )
+  (ensures (
+    let (_, tr_out) = send_authenticated #a comm_keys_ids sender receiver payload tr in
+    trace_invariant tr_out
+  ))
+  [SMTPat (trace_invariant #invs tr);
+   SMTPat (send_authenticated comm_keys_ids sender receiver payload tr);
+   SMTPat (apply_core_comm_layer_lemmas higher_layer_preds)]
+let send_authenticated_proof #invs #a tr higher_layer_preds comm_keys_ids sender receiver payload =
+    reveal_opaque (`%send_authenticated) (send_authenticated #a);
+    match send_authenticated comm_keys_ids sender receiver payload tr with
+    | (None, tr_out) -> ()
+    | (Some _, tr_out) -> (
+      let (Some sk_sender, tr') = get_private_key sender comm_keys_ids.private_keys (LongTermSigKey comm_layer_sign_tag) tr in
+      let (nonce,  tr') = mk_rand SigNonce (long_term_key_label sender) 32 tr' in
+      higher_layer_preds.send_auth_later tr tr' sender payload;
+      let payload_bytes = serialize a payload in
+      let ((), tr') = trigger_event sender (CommAuthSendMsg sender payload_bytes) tr' in
+      assert(event_triggered tr' sender (CommAuthSendMsg sender payload_bytes));
+      sign_message_proof #invs.crypto_invs #a tr' sender receiver payload None sk_sender nonce;
+      let msg_signed = sign_message #a sender receiver payload None sk_sender nonce in
+      let (msg_id, tr') = send_msg msg_signed tr' in
+      assert(tr_out == tr');
+      assert(trace_invariant tr_out);
+      ()
+    )
+
+
+#push-options "--z3rlimit 50"
+val verify_message_proof:
+  {|cinvs:crypto_invariants|} ->
+  #a:Type -> {| ps:parseable_serializeable bytes a |} ->
+  tr:trace ->
+  sender:principal -> receiver:principal ->
+  msg_bytes:bytes -> sk_receiver_opt:option bytes ->
+  vk_sender:bytes ->
+  Lemma
+  (requires
+    has_communication_layer_crypto_predicates /\
+    is_publishable tr msg_bytes /\
+    is_public_key_for tr vk_sender (LongTermSigKey comm_layer_sign_tag) sender
+  )
+  (ensures (
+    reveal_opaque (`%verify_message) (verify_message #a);
+    match verify_message #a receiver msg_bytes sk_receiver_opt vk_sender with
+    | Some cm -> (
+      let Some msg_signed_t = parse com_message_t msg_bytes in
+      let SigMessage msg_signed = msg_signed_t in
+      let Some sign_input = parse signature_input #parseable_serializeable_bytes_signature_input msg_signed.msg in
+      let payload_bytes = serialize a cm.payload in
+      is_publishable tr payload_bytes /\
+      is_well_formed a (is_publishable tr) cm.payload /\
```

Comment:
I think these two lines are equivalent (by definition of `is_well_formed`), one of them could be removed.

Reply:
Done.